crates/trie/sparse/src/state.rs

```diff
@@ -117,23 +111,51 @@ impl SparseStateTrie {
         &self,
         proof: &mut Peekable<I>,
     ) -> SparseStateTrieResult<Option<TrieNode>> {
-        let mut proof = proof.into_iter().peekable();
-
         // Validate root node.
         let Some((path, node)) = proof.next() else { return Ok(None) };
+
+        // Decode root node.
+        let root_node = TrieNode::decode(&mut &node[..])?;
+
+        // Root node path should be empty.
         if !path.is_empty() {
-            return Err(SparseStateTrieError::InvalidRootNode { path, node })
+            return Err(SparseStateTrieError::InvalidRootNodePath {
+                path,
+                node: Box::new(root_node),
+            })
         }
 
-        // Decode root node and perform sanity check.
-        let root_node = TrieNode::decode(&mut &node[..])?;
-        if matches!(root_node, TrieNode::EmptyRoot) && proof.peek().is_some() {
-            return Err(SparseStateTrieError::InvalidRootNode { path, node })
+        // No other nodes should be present if the root node is empty.
+        if (matches!(root_node, TrieNode::EmptyRoot) && proof.peek().is_some()) {
+            return Err(SparseStateTrieError::UnexpectedNode { path, node: Box::new(root_node) })
         }
 
         Ok(Some(root_node))
     }
 
+    /// Reveals the proof nodes in the provided trie.
+    fn reveal_proof<I: Iterator<Item = (Nibbles, Bytes)>>(
+        trie: &mut RevealedSparseTrie,
+        mut proof: Peekable<I>,
+    ) -> SparseStateTrieResult<()> {
+        // Reveal the remaining proof nodes.
+        while let Some((path, bytes)) = proof.next() {
+            let node = TrieNode::decode(&mut &bytes[..])?;
+
+            // Check that the last node of the proof is a leaf node.
+            if proof.peek().is_none() && !matches!(node, TrieNode::Leaf(_)) {
+                return Err(SparseStateTrieError::ProofWithoutLeafNode {
+                    path,
+                    node: Box::new(node),
+                })
+            }
```

Comment:
it is fine to have a proof without a leaf node. if the account has not been created yet, we want to reveal all nodes leading to the path